src/miroil/input_device.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright © 2016-2021 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <miroil/input_device.h>
+
+#include "mir/flags.h"
+#include "mir/input/device.h"
+#include "mir/input/mir_keyboard_config.h"
+#include "mir/input/parameter_keymap.h"
+#include "mir/version.h"
+
+miroil::InputDevice::InputDevice(std::shared_ptr<mir::input::Device> const& device)
+: device(device)
+{
+}
+
+miroil::InputDevice::InputDevice(InputDevice const& ) = default;
+
+miroil::InputDevice::InputDevice() = default;
+
+miroil::InputDevice::~InputDevice() = default;
+
+bool miroil::InputDevice::operator==(InputDevice const& other)
+{
+    return device == other.device;
+}
+        
+void miroil::InputDevice::apply_keymap(std::string const& layout, std::string const& variant)
+{
+    using namespace mir::input;
+
+    if (auto const kbd_conf = device->keyboard_configuration())
+    {
+        // TODO work out how to support this functionality on OSK etc.
+#if MIR_SERVER_VERSION >= MIR_VERSION_NUMBER(2, 5, 0)
```

Comment:
Is this check required?

Reply:
It makes for easier syncing with a mir-1.8 codebase